<img align="right" width="200" height="200" src="https://static.wikia.nocookie.net/lego/images/2/23/PPG_logo.png"> 

# <span style="color:#EF7C8E">Weekend Project - GPU EDA - Python </span>
 


## <span style="color:#55BB99"> Group: Power Puff Girls</span>
### <span style="color:#726060"> Members: Wejdan Al-Ahmadi - Sara Aldubaie</span>
<hr style="border:2px solid gray"> </hr>

In [1]:
#pip install dash-bootstrap-components

In [2]:
# Imports
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns 
import plotly.express as px
import plotly.graph_objects as go
from jupyter_dash import JupyterDash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output
import plotly.graph_objects as go
import random
import dash
import json
from dash.dependencies import Input, Output
from dash.exceptions import PreventUpdate
import dash_bootstrap_components as dbc
from datetime import datetime


In [3]:
# reading the file 
gpu = pd.read_csv('./data/All_GPUs.csv')

In [4]:
# view all columns in the dataset 
with pd.option_context('display.max_rows', None, 'display.max_columns', None): 
    display(gpu.head(30))

Architecture Best_Resolution Boost_Clock Core_Speed  \
0                  Tesla G92b              NaN         NaN    738 MHz   
1                     R600 XT       1366 x 768         NaN       \n-    
2                    R600 PRO       1366 x 768         NaN       \n-    
3                       RV630       1024 x 768         NaN       \n-    
4                       RV630       1024 x 768         NaN       \n-    
5                       RV630       1024 x 768         NaN       \n-    
6               R700 RV790 XT      1920 x 1080         NaN    870 MHz   
7                     R600 GT       1024 x 768         NaN       \n-    
8              Pitcairn XT GL      1920 x 1080         NaN       \n-    
9                       RV100              NaN         NaN       \n-    
10                  NV28GL A2              NaN         NaN       \n-    
11                Fermi GF110      1920 x 1080         NaN    650 MHz   
12               Kepler GK110              NaN         NaN    705 MHz   
13               Kepler GK110      2560 x 1600         NaN    706 MHz   
14                      RV200              NaN         NaN       \n-    
15  GCN 1.1 Oland XT + Kaveri       1600 x 900   1100 MHz    1050 MHz   
16            Kepler GK104 x2              NaN         NaN       \n-    
17               Kepler GK110              NaN         NaN    732 MHz   
18              Skylake GT3-E       1366 x 768   1000 MHz     300 MHz   
19                       R360       1024 x 768         NaN       \n-    
20                Fermi GF100              NaN         NaN    575 MHz   
21                Fermi GF100              NaN         NaN    575 MHz   
22                    NV25 A2              NaN         NaN       \n-    
23                  NV28GL A2              NaN         NaN       \n-    
24              Skylake GT3-E       1366 x 768   1000 MHz     300 MHz   
25                       NV28              NaN         NaN       \n-    
26                Fermi GF100              NaN         NaN    575 MHz   
27              Skylake GT3-E       1366 x 768   1050 MHz     300 MHz   
28              Skylake GT3-E       1366 x 768    950 MHz     300 MHz   
29                Haswell GT3       1366 x 768   1100 MHz     200 MHz   

    DVI_Connection Dedicated Direct_X  DisplayPort_Connection  \
0              2.0       Yes  DX 10.0                     NaN   
1              2.0       Yes    DX 10                     NaN   
2              2.0       Yes    DX 10                     NaN   
3              2.0       Yes    DX 10                     NaN   
4              2.0       Yes    DX 10                     NaN   
5              2.0       Yes    DX 10                     NaN   
6              1.0       Yes  DX 10.1                     NaN   
7              2.0       Yes    DX 10                     NaN   
8              0.0       Yes  DX 11.2                     NaN   
9              NaN       Yes     DX 7                     NaN   
10             2.0       Yes   DX 8.1                     NaN   
11             2.0       Yes  DX 12.0                     0.0   
12             0.0       Yes  DX 12.0                     NaN   
13             0.0       Yes  DX 12.0                     NaN   
14             NaN       Yes     DX 7                     NaN   
15             1.0       Yes  DX 12.0                     0.0   
16             0.0       Yes  DX 11.1                     NaN   
17             0.0       Yes  DX 12.0                     NaN   
18             NaN        No  DX 12.1                     NaN   
19             1.0       Yes     DX 9                     0.0   
20             1.0       Yes  DX 12.0                     NaN   
21             1.0       Yes  DX 12.0                     NaN   
22             2.0       Yes   DX 8.1                     NaN   
23             2.0       Yes   DX 8.1                     NaN   
24             NaN        No  DX 12.1                     NaN   
25             2.0       Yes   DX 8.1                     NaN   

In [5]:
# number of rows and columns 
gpu.shape

(3406, 34)

In [6]:
# number of null values in each column 
gpu.isnull().sum()

Architecture                62
Best_Resolution            642
Boost_Clock               1960
Core_Speed                   0
DVI_Connection             750
Dedicated                   14
Direct_X                     6
DisplayPort_Connection    2549
HDMI_Connection            763
Integrated                  14
L2_Cache                     0
Manufacturer                 0
Max_Power                  625
Memory                     420
Memory_Bandwidth           121
Memory_Bus                  62
Memory_Speed               105
Memory_Type                 56
Name                         0
Notebook_GPU                 0
Open_GL                     40
PSU                       1175
Pixel_Rate                 544
Power_Connector            709
Process                    463
ROPs                       538
Release_Date                 0
Release_Price             2850
Resolution_WxH             195
SLI_Crossfire                0
Shader                     107
TMUs                       538
Texture_

In [7]:
gpu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3406 entries, 0 to 3405
Data columns (total 34 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Architecture            3344 non-null   object 
 1   Best_Resolution         2764 non-null   object 
 2   Boost_Clock             1446 non-null   object 
 3   Core_Speed              3406 non-null   object 
 4   DVI_Connection          2656 non-null   float64
 5   Dedicated               3392 non-null   object 
 6   Direct_X                3400 non-null   object 
 7   DisplayPort_Connection  857 non-null    float64
 8   HDMI_Connection         2643 non-null   float64
 9   Integrated              3392 non-null   object 
 10  L2_Cache                3406 non-null   object 
 11  Manufacturer            3406 non-null   object 
 12  Max_Power               2781 non-null   object 
 13  Memory                  2986 non-null   object 
 14  Memory_Bandwidth        3285 non-null   

## <span style="color:#EF7C8E">Data cleaning:</span>

### <span style="color:#EF7C8E">1. Replace NaN with "No"</span>
* Integrated:
Is the graphics card integrated (replace NaN with "No", 14 rows)
* Dedicated:
Is the graphics card dedicated (replace NaN with "No", 14 rows)


In [8]:
# REPLACE WITH NO:
# Replace all NaN with No because these columns have either a "Yes" or "No" Value
gpu.loc[:,['Dedicated','Integrated']]=gpu.loc[:,['Dedicated','Integrated']].fillna("No")


### <span style="color:#EF7C8E">2. Replace \n, \n-, and \t with NaN</span>

* all columns 

In [9]:
# Replace all \n- and \n and \t with NaN
gpu=gpu.replace({'\n-':np.nan,'\n- ':np.nan,' \n-':np.nan,'\n':np.nan,' \n':np.nan,'\n ':np.nan,'\t':np.nan,' \t':np.nan,'\t ':np.nan})


### <span style="color:#EF7C8E">3. Replace NaN with Mode</span>

* Best_Resolution:
The best Screensize for reducing bottleneck to a minimum. (replace with mode: 642 rows)
* Direct_X:
What Direct X can the graphics card support (replace NaN with mode, 6 rows) 
* Memory:
The memory of graphics card (replace with mode: 420 rows)
* Memory_Bus:
The memory bus of graphics card (replace with mode or ffill: 62 rows)
* Memory_Type:
The memory type of the graphics card(replace with mode: 56 rows)
* Open_GL: 
What Open GL can the graphics card support.(replace with mode: 40 rows)
* Resolution_WxH:
Maximum supported resolution. (replace with mode: 195 rows)

In [10]:
# REPLACE WITH MODE: 

# 1- Best_Resolution: 
gpu["Best_Resolution"].fillna(gpu["Best_Resolution"].mode()[0], inplace=True)
# 2- Direct_X
gpu["Direct_X"].fillna(gpu["Direct_X"].mode()[0], inplace=True)
# 3- Memory
gpu["Memory"].fillna(gpu["Memory"].mode()[0], inplace=True)
# 4- Memory_Bus
gpu["Memory_Bus"].fillna(gpu["Memory_Bus"].mode()[0], inplace=True)
# 5- Memory_Type
gpu["Memory_Type"].fillna(gpu["Memory_Type"].mode()[0], inplace=True)
# 6- Open_GL
gpu["Open_GL"].fillna(gpu["Open_GL"].mode()[0], inplace=True)
# 7- Resolution_WxH
gpu["Resolution_WxH"].fillna(gpu["Resolution_WxH"].mode()[0], inplace=True)
# 8- Max_Power
gpu["Max_Power"].fillna(method="ffill", inplace=True)
# 9- Core_Speed
gpu["Core_Speed"].fillna(method="ffill", inplace=True)

### <span style="color:#EF7C8E">4. Separate Column</span>
* Name: 
Name of the GPU. (divide first part as line, second specs/line-specifications)


In [11]:

# Separate the Name into two new columns = Line and Line Description
gpu[["Line", "Line_Description"]] = gpu["Name"].str.split(" ", 1, expand=True)
# this resulted in one extra NaN so I dropped that row
gpu.dropna(subset=['Line_Description'], how='all', inplace=True)

### <span style="color:#EF7C8E">5. Date</span>
* Release_Date:
The GPU release date. (remove "\n" and "\n " and "\t" from the end, and "\nUnknown" to NaN and then replace with ffill and then convert to date format)

In [12]:
# DATE CLEANUP 

# to remove the white spaces
gpu['Release_Date']=gpu['Release_Date'].map(lambda x: x.strip())

# Replace all "Unknown Release Date" with NaN
gpu=gpu.replace({'Unknown Release Date':np.nan}) # there were 28 unknowns
# Drop the NaNs there was 28 so it wouldnt affect that much
gpu.dropna(subset=['Release_Date'], how='all', inplace=True)
# Convert to Datetime format
gpu['Release_Date']=gpu['Release_Date'].map(lambda x: datetime.strptime(x, '%d-%b-%Y'))


### <span style="color:#EF7C8E">6. Drop Rows</span>
* Architecture: 
The GPU architecture used on the graphics card. (drop rows with null: 62 rows)

In [13]:
#---------------------------------- drop rows ----------------------------------#

# drop the rows with empty value in column (Architecture) 
gpu.dropna(axis = 'index', how = 'all', subset = ['Architecture'], inplace = True)

### <span style="color:#EF7C8E">7. Replace Values with Null</span>

* Power Connector: is needed to run the graphics card(convert none to null)

In [14]:
#---------------------------------- Replace values to null ----------------------------------#

# list with values needed to be replaced
none_val = ['None ', 'None  ']
# replace 'None ' and 'None  ' values with null  
gpu['Power_Connector'] = gpu['Power_Connector'].replace(none_val, np.NaN)

### <span style="color:#EF7C8E">8. Replace NaN Values With Zero</span>

* DVI_Connection: How many DVI connections does the graphics card have (replace NaN with 0, 750 rows)
* DisplayPort_Connection: How many DisplayPort connections does the graphics card have (replace NaN with 0, 2549 rows)
* HDMI_Connection: How many HDMI connections does the graphics card have (replace NaN with 0, 763 rows)
* ROPs: How many Raster Operation Pipelines does the GPU have (replace NaN with 0, 538 rows) 
* Shader: What Shader Model can the graphics card support.(replace NaN with 0, 107 rows)
* TMUs: How many texture mapping units does the GPU have.(replace NaN with 0, 538 rows)
* VGA_Connection: How many VGA connections does GPU have.(replace NaN with 0, 758 rows)

In [15]:
#---------------------------------- Replace NaN values with zero ----------------------------------#

# list with all columns needed to replace Null value to zero 
to_zero = ['DVI_Connection', 'DisplayPort_Connection', 'HDMI_Connection','ROPs', 'Shader', 'TMUs', 'VGA_Connection']
# replace null value with zero 
gpu.update(gpu[to_zero].fillna(0))

In [16]:
gpu['Power_Connector'].value_counts()

1x 6-pin                  334
2x 6-pin                  251
1x 6-pin + 1x 8-pin       200
1x 8-pin                  150
2x 8-pin                   81
3x 8-pin                    8
4x 6-pin                    5
2x 6-pin                    2
1x 6-pin + 1x 8-pin\t       1
Name: Power_Connector, dtype: int64

### <span style="color:#EF7C8E">9. Filling Null Values Using Forward Filling</span>

* Texture_Rate: How many texture pixels per second can the GPU produce.(replace NaN with ffill, 544 rows)
* Process: What nm technology does the GPU use. (replace with ffill: 463 rows)
* Pixel_Rate: The number of pixels per second the GPU can produce.(replace with ffill/mode: 544 rows)
* PSU: What Power Supply is needed to run the graphics card.(replace with ffill: 1175 rows)
* Memory_Speed: The memory speed of the graphic card's RAM(ffill method, 105 rows)
* Memory_Bandwidth: The memory bandwidth (ffill method, 121 rows)
* Power_Connector: Power Connector is needed to run the graphics card(convert none to null and ffill 2,363 rows)
* Boost_Clock: the MHz speed of the graphics card (ffill method, 1960 rows)
    

In [17]:
#---------------------------------- filling null values using forward filling -----------------------------------#

# list with all columns needed to fill null using forward filling method
ffill_col = ['Texture_Rate', 'Process', 'Pixel_Rate', 'PSU', 'Memory_Speed', 'Memory_Bandwidth', 'Power_Connector', 'Boost_Clock']

# fill null value using forward filling method
gpu.update(gpu[ffill_col].fillna(method="ffill"))

### <span style="color:#EF7C8E">10. Filling Null Values Using Mode</span>

* Power_Connector: Power Connector is needed to run the graphics card(convert none to null and ffill 2,363 rows)
* Boost_Clock: the MHz speed of the graphics card (ffill method, 1960 rows)

In [18]:
#---------------------------------- filling null values using mode -----------------------------------#
# finding mode value 
Power_Connector_mode = gpu['Power_Connector'].mode()
Boost_Clock_mode = gpu['Boost_Clock'].mode()

# fill null value using mode
gpu['Power_Connector'] = gpu['Power_Connector'].fillna(Power_Connector_mode)
gpu['Boost_Clock'] = gpu['Boost_Clock'].fillna(Boost_Clock_mode)

### <span style="color:#EF7C8E">11. Filling Null Values Using Backward Filling</span>

* Power_Connector: Power Connector is needed to run the graphics card(convert none to null and ffill 2,363 rows)
* Boost_Clock: the MHz speed of the graphics card (ffill method, 1960 rows)

In [19]:
#---------------------------------- filling null values using backward filling -----------------------------------#
# list with all columns needed to fill null using backward filling method
bfill_col = ['Power_Connector', 'Boost_Clock']

# fill null value using backward filling method
gpu.update(gpu[bfill_col].fillna(method="bfill"))

### <span style="color:#EF7C8E">12. Filling Null in (Release_Price) column </span>

* Release_Price: The GPU price at time of release. (remove the '$', sub categorize it based on a metric/column and then fill the price with the mode of that metric/column *last step because its based on other columns) 

In [20]:
#---------------------------------- price column -----------------------------------#

# removing '$' from value and any other characters if exists
gpu['Release_Price'] = gpu['Release_Price'].replace({'\$':''}, regex = True)

# convert column type to float 
gpu['Release_Price'] = gpu['Release_Price'].astype(float)

# fill prices based on memory  
gpu['Release_Price'] = gpu['Release_Price'].fillna(round(gpu.groupby('Memory')['Release_Price'].transform('mean'),2))

# fill prices based on memory type
gpu['Release_Price'] = gpu['Release_Price'].fillna(round(gpu.groupby('Memory_Type')['Release_Price'].transform('mean'),2))


In [21]:
gpu.isnull().sum()

Architecture              0
Best_Resolution           0
Boost_Clock               0
Core_Speed                0
DVI_Connection            0
Dedicated                 0
Direct_X                  0
DisplayPort_Connection    0
HDMI_Connection           0
Integrated                0
L2_Cache                  0
Manufacturer              0
Max_Power                 0
Memory                    0
Memory_Bandwidth          0
Memory_Bus                0
Memory_Speed              0
Memory_Type               0
Name                      0
Notebook_GPU              0
Open_GL                   0
PSU                       0
Pixel_Rate                0
Power_Connector           0
Process                   0
ROPs                      0
Release_Date              0
Release_Price             0
Resolution_WxH            0
SLI_Crossfire             0
Shader                    0
TMUs                      0
Texture_Rate              0
VGA_Connection            0
Line                      0
Line_Description    

In [22]:
# view all columns in the dataset 
with pd.option_context('display.max_rows', None, 'display.max_columns', None): 
    display(gpu.head())

Architecture Best_Resolution Boost_Clock Core_Speed  DVI_Connection  \
0  Tesla G92b      1920 x 1080   1000 MHz     738 MHz             2.0   
1     R600 XT       1366 x 768   1100 MHz     738 MHz             2.0   
2    R600 PRO       1366 x 768   1100 MHz     738 MHz             2.0   
3       RV630       1024 x 768   1100 MHz     738 MHz             2.0   
4       RV630       1024 x 768   1100 MHz     738 MHz             2.0   

  Dedicated Direct_X  DisplayPort_Connection  HDMI_Connection Integrated  \
0       Yes  DX 10.0                     0.0              0.0         No   
1       Yes    DX 10                     0.0              0.0         No   
2       Yes    DX 10                     0.0              0.0         No   
3       Yes    DX 10                     0.0              0.0         No   
4       Yes    DX 10                     0.0              0.0         No   

  L2_Cache Manufacturer  Max_Power    Memory Memory_Bandwidth Memory_Bus  \
0      0KB       Nvidia  141 Watts  1024 MB          64GB/sec   256 Bit    
1      0KB          AMD  215 Watts   512 MB         106GB/sec   512 Bit    
2      0KB          AMD  200 Watts   512 MB        51.2GB/sec   256 Bit    
3      0KB          AMD  200 Watts   256 MB        36.8GB/sec   128 Bit    
4      0KB          AMD   45 Watts   256 MB        22.4GB/sec   128 Bit    

  Memory_Speed Memory_Type                               Name Notebook_GPU  \
0     1000 MHz       GDDR3                    GeForce GTS 150           No   
1      828 MHz       GDDR3            Radeon HD 2900 XT 512MB           No   
2      800 MHz       GDDR3                 Radeon HD 2900 Pro           No   
3     1150 MHz       GDDR4  Radeon HD 2600 XT Diamond Edition           No   
4      700 MHz       GDDR3                  Radeon HD 2600 XT           No   

   Open_GL                 PSU   Pixel_Rate Power_Connector Process ROPs  \
0      3.3  450 Watt & 38 Amps  12 GPixel/s       1x 6-pin     55nm   16   
1      3.1  550 Watt & 35 Amps  12 GPixel/s       2x 6-pin     80nm   16   
2      3.1  550 Watt & 35 Amps  10 GPixel/s       2x 6-pin     80nm   16   
3      3.3  550 Watt & 35 Amps   3 GPixel/s       2x 6-pin     65nm    4   
4      3.1  400 Watt & 25 Amps   3 GPixel/s       2x 6-pin     65nm    4   

  Release_Date  Release_Price Resolution_WxH SLI_Crossfire  Shader  TMUs  \
0   2009-03-01         196.40      2560x1600           Yes     4.0  64.0   
1   2007-05-14         145.45      2560x1600           Yes     4.0  16.0   
2   2007-12-07         145.45      2560x1600           Yes     4.0  16.0   
3   2007-07-01         162.43      2560x1600           Yes     4.0   8.0   
4   2007-06-28         201.89      2560x1600           Yes     4.0   8.0   

  Texture_Rate  VGA_Connection     Line            Line_Description  
0  47 GTexel/s             0.0  GeForce                     GTS 150  
1  12 GTexel/s             0.0   Radeon            HD 2900 XT 512MB  
2  10 GTexel/s             0.0   Radeon                 HD 2900 Pro  
3   7 GTexel/s             0.0   Radeon  HD 2600 XT Diamond Edition  
4   6 GTexel/s             0.0   Radeon                  HD 2600 XT

## <span style="color:#EF7C8E">Initializing New Columns For Dash Plots:</span>

In [23]:
# To plot Boost Clock as number in order
gpu['Boost_Clock_N']=gpu['Boost_Clock'].map(lambda x: x.replace(' MHz', ''))
gpu['Boost_Clock_N']=gpu['Boost_Clock_N'].astype(int)

# To plot Memory as number in order
gpu['Memory_N']=gpu['Memory'].map(lambda x: x.replace(' MB', ''))
gpu['Memory_N']=gpu['Memory_N'].astype(int)

# To plot Core Speed as number in order
gpu['Core_Speed_N']=gpu['Core_Speed'].map(lambda x: x.replace(' MHz', ''))
gpu['Core_Speed_N']=gpu['Core_Speed_N'].astype(int)

# To plot Memory Speed as number in order
gpu['Memory_Speed_N']=gpu['Memory_Speed'].map(lambda x: x.replace(' MHz', ''))
gpu['Memory_Speed_N']=gpu['Memory_Speed_N'].astype(int)

# To plot Max Power as number in order
gpu['Max_Power_N']=gpu['Max_Power'].map(lambda x: x.replace(' Watts', ''))
gpu['Max_Power_N']=gpu['Max_Power_N'].astype(int)

manufacturers = ["AMD", "Intel", 'Nvidia','ATI', "All"]
# Add year column to filter by year
gpu['year'] = pd.DatetimeIndex(gpu['Release_Date']).year
ddm_labels={'Boost Clock':'Boost_Clock_N',
                    'Memory':'Memory_N',
                    'Core Speed':'Core_Speed_N',
                    'Memory Speed':'Memory_Speed_N',
                    'Max Power':'Max_Power_N'}

## <span style="color:#EF7C8E">Dash App:</span>

In [24]:
app = JupyterDash(__name__, external_stylesheets=[dbc.themes.FLATLY])

SIDEBAR_STYLE = {
    "position": "fixed",
    "top": 0,
    "left": 0,
    "bottom": 0,
    "width": "16rem",
    "padding": "2rem 1rem",
    "background-color": "#f8f9fa",
}

# the styles for the main content position it to the right of the sidebar and
# add some padding.
CONTENT_STYLE = {
    "margin-left": "18rem",
    "margin-right": "2rem",
    "padding": "2rem 1rem",
}

sidebar = html.Div(
    [
        html.Img(src='https://static.wikia.nocookie.net/lego/images/2/23/PPG_logo.png', 
                 width="220"),
        #html.H3("Plotly Dash", className="display-4"),
        html.Hr(),
        html.P(
            "Power Puff Girls", className="lead"
        ),
        dbc.Nav(
            [
                dbc.NavLink("About Dataset", href="/", active="exact"),
                dbc.NavLink("Percentages", href="/page-1", active="exact"),
                dbc.NavLink("Manufacturer Prices", href="/page-2", active="exact"),
                dbc.NavLink("Number of GPUs", href="/page-3", active="exact"),
                dbc.NavLink("Specifications vs Line", href="/page-4", active="exact"),
            ],
            vertical=True,
            pills=True,
        ),
    ],
    style=SIDEBAR_STYLE,
)

content = html.Div(id="page-content", style=CONTENT_STYLE)

app.layout = html.Div([
    html.Div([dcc.Location(id="url"), sidebar, content]),
    
])
@app.callback(Output("page-content", "children"), 
              [Input("url", "pathname")])
def render_page_content(pathname):
    if pathname == "/":
        return  (
            # About dataset div
            html.Div([
                html.Center([
                    # title
                    html.H3(["About this Dataset:"]),
                    # body
                    html.Tbody(["This dataset contains detailed information about graphics cards, the source is: "]),
                    # button link
                    html.A(html.Button('Data', className='three columns'),
                    href='https://www.kaggle.com/iliassekkaf/computerparts')
                ])], 
            className="pretty_container", style={"width":"70%"}),
            # Table div
            html.Div([
                html.Center([
                    # title
                    html.H3(["Dataset Column Metadata:"]),
                    # table
                    html.Table([
                        html.Tr([html.Th('Architecture:'),
                                html.Td('The GPU architecture used on the graphics card.')]),
                        html.Tr([html.Th('Best_Resolution:'),
                                 html.Td('The best Screensize for reducing bottleneck to a minimum.')]),
                        html.Tr([html.Th('Boost_Clock:'),
                                 html.Td('The MHz speed of the graphics card.')]),
                        html.Tr([html.Th('Core_Speed:'),
                                 html.Td('The speed of the graphics card.')]),
                        html.Tr([html.Th('DVI_Connection:'),
                                 html.Td('How many DVI connections does the graphics card have.')]),
                        html.Tr([html.Th('Dedicated:'),
                                 html.Td('Is the graphics card dedicated.')]),
                        html.Tr([html.Th('Direct_X:'),
                                 html.Td('What Direct X can the graphics card support.')]),
                        html.Tr([html.Th('DisplayPort_Connection:'),
                                 html.Td('How many DisplayPort connections does the graphics card have.')]),
                        html.Tr([html.Th('HDMI_Connection:'),
                                 html.Td('How many HDMI connections does the graphics card have.')]),
                        html.Tr([html.Th('Integrated:'),
                                 html.Td('Is the graphics card integrated.')]),
                        html.Tr([html.Th('L2_Cache:'),
                                 html.Td('The L2 Cache of graphics card.')]),
                        html.Tr([html.Th('Manufacturer:'),
                                 html.Td('GPU manufacturer.')]),
                        html.Tr([html.Th('Max_Power:'),
                                 html.Td('What power in watts is required to run the graphics card.')]),
                        html.Tr([html.Th('Memory:'),
                                 html.Td('The memory of graphics card.')]),
                        html.Tr([html.Th('Memory_Bandwidth:'),
                                 html.Td('The memory bandwidth.')]),
                        html.Tr([html.Th('Memory_Bus:'),
                                 html.Td('The memory bus of graphics card.')]),
                        html.Tr([html.Th('Memory_Speed:'),
                                 html.Td("The memory speed of the graphic card's RAM.")]),
                        html.Tr([html.Th('Name:'),
                                 html.Td('Name of the GPU.')]),
                        html.Tr([html.Th('Notebook_GPU:'),
                                 html.Td('Is it for Notebooks or Desktop.')]),
                        html.Tr([html.Th('Open_GL:'),
                                 html.Td('What Open GL can the graphics card support.')]),
                        html.Tr([html.Th('PSU:'),
                                 html.Td('What Power Supply is needed to run the graphics card.')]),
                        html.Tr([html.Th('Pixel_Rate:'),
                                 html.Td('The number of pixels per second the GPU can produce.')]),
                        html.Tr([html.Th('Power_Connector:'),
                                 html.Td('Power Connector is needed to run the graphics card.')]),
                        html.Tr([html.Th('Process:'),
                                 html.Td('What nm technology does the GPU use.')]),
                        html.Tr([html.Th('ROPs:'),
                                 html.Td('How many Raster Operation Pipelines does the GPU have.')]),
                        html.Tr([html.Th('Release_Date:'),
                                 html.Td('The GPU release date.')]),
                        html.Tr([html.Th('Release_Price:'),
                                 html.Td('The GPU price at time of release.')]),
                        html.Tr([html.Th('Resolution_WxH:'),
                                 html.Td('Maximum supported resolution.')]),
                        html.Tr([html.Th('SLI_Crossfire:'),
                                 html.Td('Does the GPU support SLI or Crossfire Technology.')]),
                        html.Tr([html.Th('Shader:'),
                                 html.Td('What Shader Model can the graphics card support.')]),
                        html.Tr([html.Th('TMUs:'),
                                 html.Td('How many texture mapping units does the GPU have.')]),
                        html.Tr([html.Th('Texture_Rate:'),
                                 html.Td('How many texture pixels per second can the GPU produce.')]),
                        html.Tr([html.Th('VGA_Connection:'),
                                 html.Td('How many VGA connections does GPU have.')])
                
        ],className="styled-table")])], className="pretty_container", style={"width":"70%"})
 

        )
    
    elif pathname == "/page-1":
        return  ((html.Div([dcc.Dropdown( id = 'pie_ddm',
                    options=[
                        {'label': 'DVI Connections', 'value': 'DVI_Connection'},
                        {'label': 'Dedicated', 'value': 'Dedicated'},
                        {'label': 'Display Port Connection', 'value': 'DisplayPort_Connection'},
                        {'label': 'HDMI Connection', 'value': 'HDMI_Connection'},
                        {'label': 'Integrated', 'value': 'Integrated'},
                        {'label': 'Manufacturer', 'value': 'Manufacturer'},
                        {'label': 'Memory Type', 'value': 'Memory_Type'},
                        {'label': 'Notebook GPU', 'value': 'Notebook_GPU'},
                        {'label': 'San Power Connector', 'value': 'Power_Connector'},
                        {'label': 'Process', 'value': 'Process'},
                        {'label': 'SLI or Crossfire', 'value': 'SLI_Crossfire'},
                        {'label': 'Shader', 'value': 'Shader'},
                        {'label': 'VGA_Connection', 'value': 'VGA_Connection'},
                    ],
                    value='DVI_Connection',
                    searchable=True),
                 
                 html.Br([]),
                 
                 dcc.Graph(id="pie_plot")], className="pretty_container"))
                )
    
    elif pathname == "/page-2":
        manufacturers = ["AMD", "Intel", 'Nvidia','ATI', "All"]
        return (html.Div([
                html.Label(["Price:"]),
                 dcc.RangeSlider(
                    id='price_slider',
                    min=gpu['Release_Price'].min() ,
                    max=gpu['Release_Price'].max(),
                    step=1, 
                    value=[0, gpu['Release_Price'].max()]
                ),
                html.Label(["Manufacturer:",
                    dcc.RadioItems(
                    id="manufacturer_r",
                    options=[{'label':x, "value":x} for x in manufacturers],
                    value='All',
                    labelStyle = {'display': 'inline', 'cursor': 'pointer', 'margin':'10px'},
                                inputStyle={"margin-right": "10px"})  
                        ]),
                ],className="pretty_container" ),
                html.Div([ dcc.Graph(id="d3")], className="pretty_container"))
    
    elif pathname == "/page-3":
        # list for radio buttons
        manufacturers = ["AMD", "Intel", 'Nvidia','ATI', "All"]

        return ( 
            # outer div container
             html.Div([
                 # graph div
               html.Div([ 
                   dcc.Graph(id="histplot")],
                   className="pretty_container"),
                 # inputs div
               html.Div([
                   html.Label(["Year:"]),            # Label for range slider
                    dcc.RangeSlider(                 # range slider
                    id='year_range_slider',          # id
                    min=1998,                        # start year
                    max=2017,                        # end year
                    step=1,                          # steps
                    value=[2000, 2017]),             # default starting value
                   
                    html.Label(["Manufacturer:",     # label for radio button
                            dcc.RadioItems(          # radio button
                            id="manufacturer",       # id
                            options=[{'label':x, "value":x} 
                                     for x in manufacturers], # loop through and populate
                            value='All',             # default value
                            # styling
                            labelStyle = {'display': 'inline', 'cursor': 'pointer', 'margin':'10px'},
                            inputStyle={"margin-right": "10px"})])
                            ], className="pretty_container")
                 
             ], className="outer_container")
        )
    
    elif pathname == "/page-4":
        # dictionary for size radio buttons labels
        size_list={'VGA Connection':'VGA_Connection',
                   'HDMI Connection':'HDMI_Connection', 
                   'Display Port Connection':'DisplayPort_Connection', 
                   'DVI Connection':'DVI_Connection', 
                   'Shader':'Shader', 
                   'Release Price':'Release_Price', 
                   'OpenGL':'Open_GL'}
        # dictionary for drop down menu labels
        ddm_labels={'Boost Clock':'Boost_Clock_N',
                    'Memory':'Memory_N',
                    'Core Speed':'Core_Speed_N',
                    'Memory Speed':'Memory_Speed_N',
                    'Max Power':'Max_Power_N'}
        
        return (
                # outer div container
                html.Div([
                    #graph div
                    html.Div([
                        dcc.Graph(id="scatter_plot")]
                        ,className="pretty_container"),
                    # inputs div
                    html.Div([
                        html.Label(["Specifications:"], className="label"), # Label for dropdown menu
                        dcc.Dropdown(                                       # drop down menu
                            id = 'scatter_ddm',                             # id
                            options=[{'label':k, "value":v}                 # loop and populate labels/values from dictionary
                                     for k,v in ddm_labels.items()],
                            value='Boost_Clock_N',                          # default value
                            clearable=False),          
                  
                        html.Label(["Size:"],className="label"),            # Label for radio buttons
                                dcc.RadioItems(                             # radio buttons
                                    id="size",                              # id     
                                    options=[{'label':k, "value":v}         # loop and populate labels/values from dictionary
                                             for k,v in size_list.items()],
                                    value='Open_GL',                        # default value
                                    #styling
                                    labelStyle = {'display': 'inline', 'cursor': 'pointer', 'margin':'10px'},
                                    inputStyle={"margin-right": "10px"}),
                        html.Br([])], className="pretty_container"),
                    
                   
                ], className="outer_container")
            
         )
    # If the user tries to reach a different page, return a 404 message
    return dbc.Jumbotron(
        [
            html.H1("404: Not found", className="text-danger"),
            html.Hr(),
            html.P(f"The pathname {pathname} was not recognised..."),
        ]
    )

@app.callback(
    Output('pie_plot', 'figure'),
    Input('pie_ddm', 'value')
)

def pie_output(pie_column):
    names = gpu[pie_column].value_counts().index
    values = gpu[pie_column].value_counts().values
    
        
    return (
        px.pie(gpu, 
               values= values, 
               names= names,
               title = f"Based On ({pie_column})"
               )
    )

@app.callback(
    Output('d3', 'figure'),
    [Input('price_slider', 'value'),
     Input ('manufacturer_r', 'value')]
)

def _3d_output(price_range, manufacturer ):
    
    price =  gpu[(gpu['Release_Price'] >= price_range[0]) & (gpu['Release_Price'] <= price_range[1]) ]
    if(manufacturer=="All"):
        fig = px.scatter_3d(price,  
                x='Line', y='Manufacturer', z='Release_Price', color="Line", title="Manufacturer Prices")

    else: 
        fig = px.scatter_3d(price[price["Manufacturer"] == manufacturer], 
                x='Line', y='Manufacturer', z='Release_Price', color="Line", title="Manufacturer Prices")
    return ( fig
    )

@app.callback(
  Output(component_id='histplot', component_property='figure'),
    [Input(component_id='year_range_slider', component_property='value'),
    Input(component_id='manufacturer', component_property='value')]
)
def update_histplot(rang_val, manufacturer_val):
    
    # get the start and end year from the range slider
    start_year = rang_val[0]
    end_year = rang_val[1]
    
    # get the data for the range between those two years
    data = gpu[(gpu["year"]<=end_year)&(gpu["year"]>=start_year)]
    
    # if the radio button choice was all manufacturer 
    if(manufacturer_val=="All"):
        fig = px.histogram(data, x="year", color="Manufacturer",template="ggplot2")\
    #if the radio button choice had a specific manufacturer
    else: 
        fig = px.histogram(data[data["Manufacturer"]==manufacturer_val], x="year",template="ggplot2") 
    # update figure titles and axis
    fig.update_layout(yaxis_title="Count",xaxis_title="Year", title_x=0.5,title="Number of GPU's by "+manufacturer_val)
    # return figure
    return fig


@app.callback(
  Output(component_id='scatter_plot', component_property='figure'),
    [Input(component_id='scatter_ddm', component_property='value'),
    Input(component_id='size', component_property='value')]
)
def update_scatter_plot(ddm_val, size_val):
    
    # for formatting the label to add to the plot title
    for k,v in ddm_labels.items(): 
        if(ddm_val==v):
            ddm_label=k
            
    # concat the title with the dropdown label
    title_text = "Relationship between Line and "+ddm_label
    
    # draw the plot
    scatter_plot=px.scatter(
                    data_frame = gpu, 
                    x = "Line", 
                    y = gpu[ddm_val], 
                    color = "Manufacturer", 
                    size= size_val,
                    title=title_text)
    
    # for formatting the unit of the y axis
    if((ddm_val=="Boost_Clock_N")|(ddm_val=="Core_Speed_N")|(ddm_val=="Memory_Speed_N")): 
        unit=" (in MHz)"
    elif(ddm_val=="Memory_N"): 
        unit=" (in MB)"
    elif(ddm_val=="Max_Power_N"):
        unit= " (in Watts)"
        
    # update the y axis and center the plot title
    scatter_plot.update_layout(yaxis_title=ddm_label+unit, title_x=0.5)
    # return the plot
    return scatter_plot


app.config['suppress_callback_exceptions'] = True


if __name__ == '__main__':
    app.run_server(mode="external", debug = True , port = 8053)


Dash app running on http://127.0.0.1:8053/
